### PR TITLE
HPCC-14067 Avoid possible deadlock if spilling + callback called

### PR DESCRIPTION
**Pull request by jakesmith**
When the memory manager is actively trying to free memory at the
same time that a spilling stream is spilling and in the process
of deregistering it's callback a deadlock can occur - it can
block on a rows lock.
Avoid by checking if already spilt upfront in the callback and
exiting.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: thorlcr/thorutil/thmem.cpp [hunk 1]
```diff
@@ -229,6 +243,8 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
     }
     virtual bool freeBufferedRows(bool critical)
     {
+        if (spillFile) // if spilt already. NB: thread-safe, as only set by this thread and can't this freeBufferedRows() can't be called on multiple threads concurrently.
+            return false;
```
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: @richardkchapman - NB: this is the actual deadlock fix, the other changes ensure callback removed in other cases and safeguard attempt to remove >1.

File: thorlcr/thorutil/thmem.cpp [hunk 2]
```diff
@@ -229,6 +243,8 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
     }
     virtual bool freeBufferedRows(bool critical)
     {
+        if (spillFile) // if spilt already. NB: thread-safe, as only set by this thread and can't this freeBufferedRows() can't be called on multiple threads concurrently.
```
Comment (richardkchapman) [inline thorlcr/thorutil/thmem.cpp]: Not sure what the comment is trying to say but "and can't this freeBufferedRows() can't" does not make sense.

File: thorlcr/thorutil/thmem.cpp (hunk 1 above)
Comment (richardkchapman) [inline thorlcr/thorutil/thmem.cpp]: Seems a bit of a strange test - is it not possible that you need to spill again once you have spilled once?

Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: No, this stream spills all once (if at all) and reads reset from disk.

File: thorlcr/thorutil/thmem.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: will amend comment.

**Comment by jakesmith** [comment]
Closing to retarget to 5.2.8

**Pull request closed without merging**